### PR TITLE
Helper functions

### PR DESCRIPTION
**Pull request by ebenahar**

### CODE REVIEW

**Comment by ebenahar** [comment]
Separated this from https://github.com/red-hat-storage/ocs-ci/pull/69 as it got too large for a single PR

File: tests/helpers.py [hunk 1]
```diff
@@ -20,4 +41,269 @@ def create_unique_resource_name(resource_description, resource_type):
...
+    if ocp_obj.get(resource_name=resource_name):
+        logger.info(f"Deleting {resource_kind} {resource_name}")
+        assert ocp_obj.delete(resource_name=resource_name, wait=False), (
+            f"Failed to delete {resource_kind} {resource_name}"
+        )
+        if wait:
+            assert ocp_obj.wait_for_resource(
+                condition='', resource_name=resource_name, to_delete=True
+            ), f"{resource_kind} {resource_name} still exists'"
+            delete_file(TEMP_YAML)
```
Comment (keesturam) [inline tests/helpers.py]: I doubt if this should be placed within 'if wait' block? This means TEMP_YAML will be deleted only when wait = true, isn't it? is that the intention?

Comment (ebenahar) [reply tests/helpers.py]: @keesturam Good catch!

File: tests/helpers.py [hunk 2]
```diff
@@ -20,4 +41,269 @@ def create_unique_resource_name(resource_description, resource_type):
...
+    cbp_kwargs['cephblockpool_name'] = cbp_name
+    cbp_kwargs['cluster_namespace'] = project_name
+    create_resource(
+        resource_kind=CEPH_BLOCK_POOL, resource_name=cbp_name,
+        project_name=project_name, template_yaml=template_yaml,
+        wait=False, **cbp_kwargs
+    )
+
+
+def create_storageclass(
```
[the remark this answers is not in the document]
Comment (ebenahar) [reply tests/helpers.py]: @keesturam Thanks, will try to add these abilities

File: tests/helpers.py [hunk 3]
```diff
@@ -20,4 +41,269 @@ def create_unique_resource_name(resource_description, resource_type):
...
+        project_name (str): The name of the project to create
+
+    Raises:
+        AssertionError: In case of any failure
+    """
+    ocp_project = ocp.OCP(kind='namespace')
+    ocp_project.get()
+    assert ocp_project.create(
+        resource_name=project_name
+    ), f"Failed to delete project {project_name}"
```
[the remark this answers is not in the document]
Comment (ebenahar) [reply tests/helpers.py]: Done

File: tests/helpers.py [hunk 4]
```diff
@@ -20,4 +41,269 @@ def create_unique_resource_name(resource_description, resource_type):
...
+
+    Args:
+        project_name (str): The name of the project to delete
+
+    Raises:
+        AssertionError: In case of any failure
+    """
+    ocp_project = ocp.OCP(kind='namespace')
+    assert ocp_project.delete(resource_name=project_name), (
+        f"Failed to create project {project_name}"
```
[the remark this answers is not in the document]
Comment (ebenahar) [reply tests/helpers.py]: done

File: tests/helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vasukulkarni) [reply tests/helpers.py]: Please check and if its missing functionality add it here: https://github.com/red-hat-storage/ocs-ci/blob/master/ocs/rook.py#L169-L174

File: tests/helpers.py [hunk 5]
```diff
@@ -20,4 +41,269 @@ def create_unique_resource_name(resource_description, resource_type):
...
+    assert ocp_obj.create(yaml_file=TEMP_YAML), (
+        f"Failed to create {resource_kind} {resource_name}"
+    )
+    assert ocp_obj.get(resource_name=resource_name), (
+        f"Failed to create {resource_kind} {resource_name}"
+    )
+    if wait:
+        assert ocp_obj.wait_for_resource(
+            condition=desired_status, resource_name=resource_name
+        ), f"{resource_kind} {resource_name} failed to reach status Available"
```
[the remark this answers is not in the document]
Comment (ebenahar) [reply tests/helpers.py]: Done

File: tests/helpers.py [hunk 6]
```diff
@@ -20,4 +41,269 @@ def create_unique_resource_name(resource_description, resource_type):
...
+    pod_kwargs['pod_name'] = pod_name
+    pod_kwargs['pvc_name'] = pvc_name
+    pod_kwargs['cluster_namespace'] = project_name
+    create_resource(
+        POD, pod_name, project_name, template_yaml,
+        desired_status='Running', **pod_kwargs
+    )
+
+
+def create_project(project_name):
```
[the remark this answers is not in the document]
Comment (ebenahar) [reply tests/helpers.py]: done

File: tests/helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ebenahar) [reply tests/helpers.py]: @vasukulkarni AFAIU, this implementation is using the client. I prefer to use the cli

**Comment by RazTamir** [comment]
@ebenahar could you please change the call under test_add_mds_to_cluster.py?

**Comment by ebenahar** [comment]
@RazTamir done

File: resources/base_resource.py [hunk 7]
```diff
@@ -1,36 +1,39 @@
...
 from ocs.ocp import OCP
-
-from ocs import defaults
+from utility import utils
+from utility import templating
 
 log = logging.getLogger(__name__)
 
 
 class BaseOCSClass(object):
```
Comment (vasukulkarni) [inline resources/base_resource.py]: If we are going to keep only one class, I think we can just rename this as "OCS" and move here https://github.com/red-hat-storage/ocs-ci/tree/master/ocs

File: tests/test_add_mds_to_cluster.py [hunk 8]
```diff
@@ -42,8 +40,7 @@ def setup(self):
     Setting up the environment for the test
     """
     global CEPH_OBJ
-    CEPH_OBJ = cephfs.CephFileSystem(**self.fs_data)
-    CEPH_OBJ.create()
+    CEPH_OBJ = helpers.create_ocs_obj(**self.fs_data)
```
Comment (RazTamir) [inline tests/test_add_mds_to_cluster.py]: What about calling creat()?

Comment (ebenahar) [reply tests/test_add_mds_to_cluster.py]: @RazTamir  called the wrong function :)

File: resources/base_resource.py (hunk 7 above)
Comment (ebenahar) [reply resources/base_resource.py]: @vasukulkarni tend to agree

File: tests/helpers.py [hunk 9]
```diff
@@ -21,3 +26,39 @@ def create_unique_resource_name(resource_description, resource_type):
...
+    Create an instance of BaseOCSClass
+
+    Args:
+        kwargs (dict): Dictionary of the OCS resource
+
+    Returns:
+        BaseOCSClass: An instance of BaseOCSClass
+
+    """
+    return BaseOCSClass(**kwargs)
```
Comment (vasukulkarni) [inline tests/helpers.py]: why do we need this,  one can directly call OCS()

Comment (ebenahar) [reply tests/helpers.py]: done

**Comment by vasukulkarni** [comment]
LGTM, if you can also post the test results please, thanks!

File: resources/storage_class.py [hunk 10]
```diff
@@ -7,13 +7,12 @@
...
 
-
-from resources.base_resource import BaseOCSClass
+from resources.ocs import OCS
 
 log = logging.getLogger(__name__)
 
 
-class StorageClass(BaseOCSClass):
+class StorageClass(OCS):
```
Comment (RazTamir) [inline resources/storage_class.py]: @ebenahar Are we keeping it?

Comment (ebenahar) [reply resources/storage_class.py]: @RazTamir Not sure what to do we these files. Let's wait with the deletion for another PR

**Comment by ebenahar** [comment]
Executed this case. Please note the cephFS creation is with wait=False as wait_for_resource has some issue that seems to be unrelated to the change done here. Will be examined along with a test case PR.

```python
import logging

from tests import helpers
from ocs import defaults
logger = logging.getLogger(__name__)


class Test1:
    def test_1(self):
        ceph_fs_dict = defaults.CEPHFILESYSTEM_DICT.copy()
        fs_name = helpers.create_unique_resource_name('test', 'fs')
        ceph_fs_dict['metadata']['name'] = fs_name
        fs_obj = helpers.create_resource(**ceph_fs_dict, wait=False)
```


=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /home/ebenahar/git/ocs-ci/ocs-venv/bin/python
cachedir: .pytest_cache
rootdir: /home/ebenahar/git/ocs-ci, inifile: pytest.ini
plugins: ocs-ci-0.1
collected 1 item                                                                                                                                                                                                  

tests/test_1.py::Test1::test_1 
-------------------------------------------------------------------------------------------------- live log call --------------------------------------------------------------------------------------------------
00:44:12 - MainThread - resources.ocs - INFO - Adding CephFilesystem with name fs-test-0600441208
00:44:12 - MainThread - utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ebenahar/ebenahar-ocs/auth/kubeconfig create -f /tmp/CephFilesystemvl641phb -o yaml
PASSED

File: resources/storage_class.py (hunk 10 above)
Comment (RazTamir) [reply resources/storage_class.py]: ok

File: tests/test_add_mds_to_cluster.py [hunk 11]
```diff
@@ -42,8 +40,7 @@ def setup(self):
     Setting up the environment for the test
     """
     global CEPH_OBJ
-    CEPH_OBJ = cephfs.CephFileSystem(**self.fs_data)
-    CEPH_OBJ.create()
+    CEPH_OBJ = helpers.create_resource(**self.fs_data)
     assert POD.wait_for_resource(
```
Comment (RazTamir) [inline tests/test_add_mds_to_cluster.py]: so now this needs to be removed as the create_resource is waiting.
Can you try to execute this test as a verifier?

Comment (RazTamir) [reply tests/test_add_mds_to_cluster.py]: actually this is not really good to use create_resource as it needs more modification.
Missing selector so you won't be able to get only mds pods.
Just call OCS(kwargs) for now

**Comment by ebenahar** [comment]
@RazTamir test_add_mds_to_cluster is still unstable. After apply, verify_fs_exist() fails as the pod fails to get to status Running. Will continue examining this.

**Comment by RazTamir** [comment]
@ebenahar your environment is probably dirty from previous executions.

I executed with your branch:
================================================================================== short test summary info ==================================================================================
PASSED tests/test_add_mds_to_cluster.py::TestCephFilesystemCreation::test_cephfilesystem_creation

Merging this

**Pull request merged**